### PR TITLE
:sparkles: Ensure logging is uniform

### PR DESCRIPTION
**Pull request by DirectXMan12**
This ensures that logging uses a uniform set of tags and "style" for the tags.  It also adds a helper that prevents the entire object from being logged when logging kubernetes objects.  Instead, only a few key identifier fields (`name`, `namespace`, `api version`, and `kind`, when relevant) are logged when in production (the full object is still logged in development).

### CODE REVIEW

**Comment by DirectXMan12** [comment]
Please take a look at the logging guidelines, and then I'll move them to the kb book as well.

**Comment by DirectXMan12** [comment]
cc @droot

File: TMP-LOGGING.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  instance, if you have a tag `api version`, use the term `APIVersion`
+  instead of `GroupVersion` in your message.
+
+## Logging Kubernetes Objects
+
+Kubernetes objects should be logged directly, like `log.Info("this is
+a Kubernetes object", "pod", somePod)`.  controller-runtime provides
+a special encoder for Zap that will transform Kubernetes objects into
+`name, namespace, apiVersion, kind` objects, when available and not in
+development mode.  Other logr implementations should implement similar
```
Comment (droot) [inline TMP-LOGGING.md]: This is sweet.

File: TMP-LOGGING.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+logic.
+
+## Logging Tags
+
```
Comment (droot) [inline TMP-LOGGING.md]: It will be useful to highlight what `Tags` are and how they impact the log message and what is their utility ?

Comment (DirectXMan12) [reply TMP-LOGGING.md]: replaced the word "tag" with "key-value pairs" to make things more obvious.

Comment (DirectXMan12) [reply TMP-LOGGING.md]: and expanded on things a bit.

File: TMP-LOGGING.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+Finally, you can use `V(1)` to mark a particular log line as "debug" logs:
+
+```go
+logger.V(1).Info("this is particularly verbose!", "state of the world",
+allKubernetesObjectsEverywhere)
+```
+
+While it's possible to use higher log levels, it's reccomended that you
+stick with `V(1)` or V(0)` (which is equivalent to not specifying `V`),
```
Comment (mengqiy) [inline TMP-LOGGING.md]: > V(0)`

nit: a missing "`"

File: TMP-LOGGING.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+conveys that an error occurred in some capacity, but that no actual
+`error` object was involved.
+
+## Logging messages
+
+- Don't put variable content in your messages -- use key-value pairs for
+  that. Never use `fmt.Sprintf` in your message.
+
+- Try to match the terminology in your messages with your key-value pairs
+  -- for instance, if you have a key-value pairs `api version`, use the
```
Comment (mengqiy) [inline TMP-LOGGING.md]: I can't follow this example.

Is `api version` the key in the K-V pair?

And it does using `APIVersion` conflicts with a later section saying
```
Use lower-case, space separated keys.  For example `object` for objects,
  `api version` for `APIVersion`
```

File: TMP-LOGGING.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+- If code works with a generic Kubernetes `runtime.Object`, use the
+  `object` key.  For specific objects, prefer the resource name as the key
+  (e.g. `pod` for `v1.Pod` objects).
+
+- For non-Kubernetes objects, the `object` key may also be used, if you
+  accept a generic interface.
+
+- When logging a raw type, log it using the `type` key, with a value of
+  `fmt.Sprintf("%T", typ)`
```
Comment (mengqiy) [inline TMP-LOGGING.md]: typo here?

File: TMP-LOGGING.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+- If there's specific context around a type, the key may be more specific,
+  but should end with `type` -- for instance, `OwnerType` should be logged
+  as `owner` in the context of `log.Error(err, "Could not get ObjectKinds
+  for OwnerType", `owner type`, fmt.Sprintf("%T"))`.  When possible, favor
```
Comment (mengqiy) [inline TMP-LOGGING.md]: > fmt.Sprintf("%T")

Need an object here.

File: TMP-LOGGING.md (hunk 2 above)
Comment (droot) [reply TMP-LOGGING.md]: Looks good.

**Comment by droot** [comment]
/lgtm

Will wait @mengqiy to take a look and approve since he had a few comments to be addressed.

**Comment by mengqiy** [comment]
It seems the bot reacted too fast :laughing: 
I'm OK to let Solly address the minor comments in a followup PR.

**Pull request merged**